File: PolynomialRegression.R
######################################Data Preprocessing#####################################

#1 importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[,2:3]

#2 dealing with missing data

# dataset$ = ifelse(is.na(dataset$),
#                   ave(dataset$, FUN = function(x) mean(x,na.rm = TRUE)),
#                   dataset$)


#3 categorical data

# dataset$State = factor(dataset$State, levels = c('New York', 'California', 'Florida'),
#                        labels = c(1,2,3))

#4 Splitting into training and testing sets
# 
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Profit, SplitRatio = 0.8)
# 
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# 
# #5 Feature Scaling
# 
# training_set[] = scale(training_set[])
# test_set[] = scale(test_set[])




################################# Data Preprocessing Completed ############################

################################## Ploynomial Regression ##################################


#1 Create a linear regressor and provide the training set
linear_reg = lm(formula = Salary ~ .,data = dataset)


#2 Create a Polynomial regressor and provide the training set
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,data = dataset)



#3 Visualizing the linear model
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),colour = 'black') +
  geom_line(aes(x = dataset$Level, y = predict(linear_reg, newdata = dataset)),colour = 'red') +
  ggtitle('Salary VS Level (Linear Regression)')
xlab('Level')
ylab('Salary')

#3 Visualizing the Polynomial model
library(ggplot2)
X_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),colour = 'black') +
  geom_line(aes(x = X_grid, y = predict(poly_reg, newdata = data.frame(Level = X_grid, 
                                                                       Level2 = x_grid^2,
                                                                       Level3 = x_grid^3,
                                                                       Level4 = x_grid^4))),colour = 'red') +
  ggtitle('Salary VS Level (Polynomial Regression)')
xlab('Level')
ylab('Salary')



# Predicting a new result with Linear Regression
predict(linear_reg, data.frame(Level = 7.5))

# Predicting a new result with Polynomial Regression
predict(poly_reg, data.frame(Level = 7.5,
                             Level2 = 7.5^2,
                             Level3 = 7.5^3,
                             Level4 = 7.5^4))


################################ PolYnomial Regression Completed######################################

File: PolynomialRegression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 30 12:09:29 2018

@author: amal
"""

########################### Data Preprocessing ###################################

#1 Importing the libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#2 importing the data set

dataset = pd.read_csv('Position_Salaries.csv')


#3 classifying dependent and independent variables

X = dataset.iloc[:,1:2].values
Y = dataset.iloc[:,2].values

"""
#4 dealing with missing values

from sklearn.preprocessing import Imputer
imp = Imputer(missing_values="NaN", strategy="mean", axis=0)
X[:,:-1] = imp.fit_transform(X[:,:-1])
"""

"""
#5 dealing with categorical data

from sklearn.preprocessing import LabelEncoder,OneHotEncoder
label_X = LabelEncoder()
X[:,-1] = label_X.fit_transform(X[:,-1])
ohe_X = OneHotEncoder(categorical_features = [3])
X = ohe_X.fit_transform(X).toarray()
"""

"""Avoiding the Dummy variable trap """
#X = X[:,1:]  # some libraries require it to be taken care manualy

"""
#6 Splitting into training and testing set

from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size = 0.2,random_state = 0)
"""

"""
#7 Feature Scaling

from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.fit(X_test)

"""


#################Data Preprocessing Ends #################################



########################## Polynomial Regression ##################################


#1 Create a linear regressor and provide the dataset
from sklearn.linear_model import LinearRegression 
Lin_regressor = LinearRegression()
Lin_regressor.fit(X, Y)


#2 Create a Polynomial regressor and update the independent variable set

from sklearn.preprocessing import PolynomialFeatures
Poly_reg = PolynomialFeatures(degree = 4)
X_poly = Poly_reg.fit_transform(X)

Lin_regressor2 = LinearRegression()
Lin_regressor2.fit(X_poly, Y)

#Visualizing the linear regression model
plt.scatter(X, Y, color = 'black')
plt.plot(X, Lin_regressor.predict(X), color = 'blue')
plt.title('Salary vs Level (Linear Model)')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

#Visualizing the Polynomial regression model
X_grid = np.arange(min(X),max(X),0.1)
X_grid = X_grid.reshape(len(X_grid),1)
plt.scatter(X, Y, color = 'black')
plt.plot(X_grid, Lin_regressor2.predict(Poly_reg.fit_transform(X_grid)), color = 'red')
plt.title('Salary vs Level (Polynomial Model)')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()


#Predicted salary of a level  7.5 employee
Lin_regressor2.predict(Poly_reg.fit_transform(7.5))
